sgt-back-end/server.ts

```diff
@@ -0,0 +1,137 @@
+  try {
+    const gradeId = Number(req.params.gradeId);
+    if (!Number.isInteger(gradeId) || gradeId <= 0) {
+      throw new ClientError(400, '"gradeId" must be a positive integer');
+    }
+    const sql = `
+      select *
+      from "grades"
+      where "gradeId" = $1;
+    `;
+    const params = [gradeId];
+    const result = await db.query(sql, params);
+    const grade = result.rows[0];
+    if (!grade) {
+      throw new ClientError(404, `Cannot find grade with "gradeId" ${gradeId}`);
+    }
+    res.json(grade);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.put('/api/grades/:gradeId', async (req, res, next) => {
+  try {
+    const gradeId = Number(req.params.gradeId);
+    if (!Number.isInteger(gradeId) || gradeId <= 0) {
+      throw new ClientError(400, '"gradeId" must be a positive integer.');
+    }
+    const { name, course, score } = req.body;
+    if (
+      !name ||
+      !course ||
+      score === undefined ||
+      score < 0 ||
+      score > 100 ||
+      !Number.isInteger(score)
+    ) {
+      throw new ClientError(400, 'Invalid grade data provided.');
+    }
+    const sql = `
+      update "grades"
+      set name = $1, course = $2, score = $3
+      where "gradeId" = $4
+      returning *;
+    `;
+    const params = [name, course, score, gradeId];
+    const result = await db.query(sql, params);
+    if (result.rows.length === 0) {
+      throw new ClientError(404, `Cannot find grade with "gradeId" ${gradeId}`);
+    }
+    res.json(result.rows[0]);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.post('/api/grades', async (req, res, next) => {
+  try {
+    const { name, course, score } = req.body;
+    if (
+      !name ||
+      !course ||
+      score === undefined ||
+      score < 0 ||
+      score > 100 ||
+      !Number.isInteger(score)
+    ) {
+      throw new ClientError(400, 'Invalid grade data provided.');
+    }
+    const sql = `
+      insert into "grades" (name, course, score)
+      values ($1, $2, $3)
+      returning *;
+    `;
+    const params = [name, course, score];
+    const result = await db.query(sql, params);
+    res.status(201).json(result.rows[0]);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.delete('/api/grades/:gradeId', async (req, res, next) => {
+  try {
+    const gradeId = Number(req.params.gradeId);
+    if (!Number.isInteger(gradeId) || gradeId <= 0) {
+      throw new ClientError(400, '"gradeId" must be a positive integer.');
+    }
+    const sql = `
+      delete from "grades"
+      where "gradeId" = $1
+      returning *;
+    `;
+    const params = [gradeId];
+    const result = await db.query(sql, params);
+    if (result.rows.length === 0) {
+      throw new ClientError(404, `Cannot find grade with "gradeId" ${gradeId}`);
+    }
+    res.status(204).end();
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.use(errorMiddleware);
+
+app.listen(8080, () => {
+  console.log('server is listening');
```

Comment:
It's usually advisable to log which port it's listening on because you'll probably forget. :-)